### DISCUSSION TITLE
Key bindings, auto resolution, better mod options...

### DISCUSSION THREAD

**Discussion by Tirlititi** [Albeoris/Memoria#978] [Releases]
![Release2024-12-04](https://github.com/user-attachments/assets/2e602772-c50f-441a-bd19-9694cbf77173)
[COMPLETE CHANGELOG HERE](https://github.com/Albeoris/Memoria/wiki/Changelog-v2024.12.04)

## Features

- Added ability to rebind the keys WASD, Pause and Help (via ``Memoria.ini``)
- Restored ``AntiAliasing`` option, only accessible via ``Memoria.ini`` (use at your own risk)
- Reduced sensitivity of UI navigation with the analog stick

## Launcher

- Auto resolution has been added
- Borderless Fullscreen mode will now always use the current resolution of the selected display
- Improved Mod options functionality and display
- Updated brazilian portuguese translation by @Felipefpl
- Fixed Excalibur II time limit tooltips (thanks @SimonAfek)

## Backend

- It is now possible to specify a condition in ``Items.csv`` for using the item
- Items not marked as "Usable" cannot be used in battles anymore: it allows to have items that can be used out of battles only ("Consumable")
- ``Min`` and ``Max`` in NCalc formulas can now be used with more than 2 arguments
- Add the possibility to use icons in ``btl2d`` messages

## Bugfixes

- Fixed a hard-lock that could occur when using the PSX-Like command menu
- Fixed mod options deactivating while using the launcher #971
- The game window will no longer be resized when Fullscreen is selected
- Various fixes regarding languages and translations
- Various other small fixes

@snouz @Tirlititi @DV666 @SamsamTS @faospark @barkermn01 @EricHu33 @Albeoris

<hr /><em>This discussion was created from the release <a href='https://github.com/Albeoris/Memoria/releases/tag/v2024.12.04'>Key bindings, auto resolution, better mod options...</a>.</em>

**Comment by moonstar207** [comment]
Hello, came across a problem on my new playthrough.  On the Festival of the Hunt part when it starts the screen goes black. I can still open the menu and pause and all that. Waited until the hunt is over and it's still black screen unfortunately. Any ideas? I tried disabling all the mods too to no avail.

**Reply by SamsamTS** [reply]
No problems with the hunt on my end. Could you provide your `Memoria.log`?

**Reply by moonstar207** [reply]
Ok here it is!
[Memoria.log](https://github.com/user-attachments/files/18063172/Memoria.log)
For reference, I have Moguri, Beatrix mod, Perfect Stats, Playstation Sounds, Garnet as Main Character, and Kupo Controller Pack installed! I had Playable character pack at the beginning but eventually disabled it.

**Reply by Tirlititi** [reply]
It might be a problem with the "Garnet is Main Character" mod. I haven't heard of that bug with that mod before, but I never read much feedback on that mod either.

The problem is that it can be complicated to fix if disabling the mod isn't enough. Maybe try to use the [Memoria Save Editor](https://forums.qhimm.com/index.php?topic=11494.0): make it so only Zidane is in the party (if I understand where you are, that's what it should be without mods) and also change a bit the game time, so the Save Editor's changes are taken into account, and also disable the mod for the duration of the hunt.

**Reply by moonstar207** [reply]
Yep, I tried it and it works. It seems the Garnet mod is the reason haha. Thanks for this! I'll get through this part then use the mod again afterwards.

EDIT: After the hunt, the award scene went off without a problem with the mod on!

**Comment by charlie-in-spain** [comment]
Hi there,

I post this issue here since maybe you guys have come up with a solution in the latest updates of Memoria to fix this.
I've created new items in HW for my mod, but I cannot select them as steal or drop items in enemies. The last item available to select is the one with ID 254 (ore). I've search in BattlePatch file wiki in case this can be added here somehow, but I see that it works like in HW: the last item is 254: https://github.com/Albeoris/Memoria/blob/main/Assembly-CSharp/Memoria/Data/Battle/RegularItem.cs

Are there any options to do this? I'd like to set a new created item as a drop for a boss later on in my mod. 

Thanks in advance!
Regards.

**Reply by Tirlititi** [reply]
Hi.
You can also use numerical IDs in ``BattlePatch.txt``

**Reply by charlie-in-spain** [reply]
Hi there! 

I've tried with "WinItems" but I can't manage to get the items I set in BattlePatch.txt. I tried with and without the "=" character:
== Battle: 742 ==
WinItems: 256 259 260 261 

== Battle: 742 ==
= WinItems: 256 259 260 261 =

I assume I should always get the item 256 if I don't change drop rates, right?

**Reply by Tirlititi** [reply]
You forget to select the enemy. It's not enough to select the battle, you must also select the enemy for patching item drops.
```txt
== Battle: 742 ==
= Enemy: 0 =
WinItems: 256 259 260 261
```

**Reply by charlie-in-spain** [reply]
Oh, I thought it was not necessary to select the enemy and that rewards were related only to the battle itself. 
Now it works fine with the enemy set. Thanks!! ;)

**Reply by charlie-in-spain** [reply]
Hi again! 

It seems that ID items is working fine for BattlePatch.txt, but not for the items I change in fields scripts. 
If I set "if ( ( ( GetItemCount(258) < 99 ) && ... " in a script, the game will get me a "Memory Earring", which is the third "key item", since it is counting 255 from "normal" items + 3. I'm not getting my new "normal" item with ID 258. 
I guess "Menu position" has nothing to do with this, since I've tried to place it after "open 255" but the issue is still happening. 

Maybe I should set something else? 

Regards!

**Reply by Tirlititi** [reply]
Hi.
Use ``GetItemCount(ConvertRegularItemId(258)) < 99`` for that kind of things. Check the documentation (in the "function list") of these conversion codes.

**Reply by charlie-in-spain** [reply]
Hi! 
Perfect, it works fine now! You are right, I hadn't checked the reference in "function list".
Thanks for the explanation!

**Comment by charlie-in-spain** [comment]
Hi there, 

I have a strange issue in my mod, I don't know since when this is happening though. 
Characters deal the same damage both in front and back rows, and also get the same damage from enemies. I wonder if this can be related to external battle scripts, but I didn't change anything of this feature. Apart from the new external scripts I've created, I've added "CommonScript.cs" from AF mod (where I added a custom battle status).
I've tried to add "DamagemodifierScript.cs" from AF into my mod, and then this works fine but Zidane's trance deals 0 damage. I don't know if this was somehow intended under some circumstances. How could I only fix the rows issue, without adding any more features?

Regards!

EDIT: It's not only Zidane's trance that deals 0 damage, but also some attacks from characters and from enemies. I can't get to understand why this happens in my mod with "DamegemodifierScript.cs" and it works fine in AF mod. If I don't add this script this doesn't happen but then I have the front/back row issue, dealing the same damage.

**Reply by Tirlititi** [reply]
Hello.
In that kind of situations, check the ``Memoria.log``: it usually gives insight on what goes wrong.
If it doesn't say anything out of the ordinary, then surely the code you use is doing something wrong. In that case, try understanding how these C# code work, to check step by step how the damage are computed. You can be helped in that debugging process by adding logging messages (that will appear in ``Memoria.log``) in the code. For example:
```csharp
Memoria.Prime.Log.Message($"[Debug Damage Scripts] Current damage parameters: AttackPower = {_v.Context.AttackPower}, DefensePower = {_v.Context.DefensePower}, Attack = {_v.Context.Attack} --> Damage = {_v.Target.HpDamage}");
```

**Reply by charlie-in-spain** [reply]
Hi Tirlititi, 

Thanks for the tip. I've tried to use the C# scripts and files from your AF mod (external, sequences, textures, etc), and after recompiling them in my mod I still get the same errors in Memoria.log file that I get with my own scripts (which compile well), so I don't know what's going on:  

_23.02.2025 03:46:46 |E| System.ArgumentException: The requested value 'Darkness' was not found.
23.02.2025 03:46:46 |E|   at System.Enum.Parse (System.Type enumType, System.String value, Boolean ignoreCase) [0x00000] in <filename unknown>:0 
23.02.2025 03:46:46 |E|   at System.Enum.Parse (System.Type enumType, System.String value) [0x00000] in <filename unknown>:0 
23.02.2025 03:46:46 |E|   at NCalc.NCalcUtility+<>c.<.cctor>b__15_1 (System.String name, NCalc.ParameterArgs args) [0x00000] in <filename unknown>:0 
23.02.2025 03:46:46 |E|   at NCalc.Domain.EvaluationVisitor.OnEvaluateParameter (System.String name, NCalc.ParameterArgs args) [0x00000] in <filename unknown>:0 
23.02.2025 03:46:46 |E|   at NCalc.Domain.EvaluationVisitor.Visit (NCalc.Domain.Identifier parameter) [0x00000] in <filename unknown>:0 
23.02.2025 03:46:46 |E|   at NCalc.Domain.Identifier.Accept (NCalc.Domain.LogicalExpressionVisitor visitor) [0x00000] in <filename unknown>:0 
23.02.2025 03:46:46 |E|   at NCalc.Domain.EvaluationVisitor+<>c__DisplayClass13_0.<Visit>b__1 () [0x00000] in <filename unknown>:0 
23.02.2025 03:46:46 |E|   at NCalc.Domain.EvaluationVisitor.Visit (NCalc.Domain.BinaryExpression expression) [0x00000] in <filename unknown>:0 
23.02.2025 03:46:46 |E|   at NCalc.Domain.BinaryExpression.Accept (NCalc.Domain.LogicalExpressionVisitor visitor) [0x00000] in <filename unknown>:0 
23.02.2025 03:46:46 |E|   at NCalc.Domain.EvaluationVisitor.Visit (NCalc.Domain.TernaryExpression expression) [0x00000] in <filename unknown>:0 
23.02.2025 03:46:46 |E|   at NCalc.Domain.TernaryExpression.Accept (NCalc.Domain.LogicalExpressionVisitor visitor) [0x00000] in <filename unknown>:0 
23.02.2025 03:46:46 |E|   at NCalc.Domain.EvaluationVisitor.Visit (NCalc.Domain.TernaryExpression expression) [0x00000] in <filename unknown>:0 
23.02.2025 03:46:46 |E|   at NCalc.Domain.TernaryExpression.Accept (NCalc.Domain.LogicalExpressionVisitor visitor) [0x00000] in <filename unknown>:0 
23.02.2025 03:46:46 |E|   at NCalc.Domain.EvaluationVisitor.Visit (NCalc.Domain.TernaryExpression expression) [0x00000] in <filename unknown>:0 
23.02.2025 03:46:46 |E|   at NCalc.Domain.TernaryExpression.Accept (NCalc.Domain.LogicalExpressionVisitor visitor) [0x00000] in <filename unknown>:0 
23.02.2025 03:46:46 |E|   at NCalc.Expression.Evaluate () [0x00000] in <filename unknown>:0 
23.02.2025 03:46:46 |E|   at Memoria.Data.SupportingAbilityFeature.TriggerOnBattleResult (.PLAYER play, .BONUS bonus, System.Collections.Generic.List`1 bonus_item, System.String when, UInt32 fleeGil) [0x00000] in <filename unknown>:0_ 

I'm not sure about what I could have done, but now I don't have the 0 damage issue with trances. The front/back row issue persists though. 


I've tried to add the line you mention in some C# scripts from my mod, so that it displays errors in Memoria.log file, but I must be doing it wrong since I get some errors in the compiler, such as "The name '_v' does not exist in the current context". Should I place it at the beginning of the script, at the end...? 

Thanks!

**Reply by charlie-in-spain** [reply]
Hi again! 

I managed to get no errors in Memoria.log file. It turns out that it was a SA from my AbilityFeatures.txt file. Once I removed it from this file, the log was not displaying any errors. The SA was working fine though... 

...However, I still have the front/back row issue. I've tried to put no scripts in my mod, so that it uses the ones from the original FF9, and then front/back row works fine. Then I guess this behaviour must be related to battle scripts somehow, but Memoria.log file doesn't display anything. Could you please put an example of where the logging message should be added in an external C# script? 

In my mod I've made some modifications of the original ones, and also new ones, but none of them affects the damage dealed in front/back row...

**Reply by charlie-in-spain** [reply]
Hi! 

I've checked all the external battle scripts in my mod. I've been placing them one by one in the scripts folder, I've compiled each time, and then I played the game to check which one could be the cause of the bug. Now I don't have it anymore, but I haven't edited any of my scripts. 
I've done this in a new mod folder I created, with all my files of the mod. I don't know if this has fixed it...

Regards!

**Comment by charlie-in-spain** [comment]
Hello! 

I've detected an issue in Memoria.ini file, I don't know if I'm doing something wrong.

"AccessMenus" is not working fine. If I put, for example, "AccessMenus = 3" and save Memoria.ini file, once the launcher is loaded I can see that this parameter has been set to 0 automatically in the file: "AccessMenus = 0". So I can't get in the menu in the original game. 
I've tried to add this parameter in Memoria.ini file of Alternate Fantasy mod. In this case, the access to the menu works at the beginning when the ATB bars are being loaded, and also when I activate the Auto-mode in battle, which has no sense. During the battle I can't access to the menu at any time. 
However, in my own mod this feature is working fine, and I've set the same parameter in Memoria.ini file. I can't understand why.

The reason why I wanted to try this feature in the original game or AF mod is because I've detected that doing this when controlling Zidane in trance and selecting a different weapon for him during the battle he gets bugged in my mod. You can select any command for him afterwards but the action is never executed, and the battle doesn't continue any more. Zidane keeps in a strange frozen position in this case: 

![zidane](https://github.com/user-attachments/assets/b75f3952-59c5-4f7f-96e5-727ce5c48c4e)

I've tried to delete the Memoria.ini file so that a new one is created when launching the game, and reinstalled Memoria patch. Do you know what could be wrong?

**Reply by Tirlititi** [reply]
I think I already fixed the problem with Zidane's T-pose that triggers in some situations. The fix will be part of the next release.

For the ``Memoria.ini`` bug, that's most likely because the options ``AccessMenus`` and ``AvailableMenus`` are intertwinned in the launcher's interface, so the launcher doesn't identify the option "dynamic ATB mode" because for the launcher it's not enough to have ``AccessMenus = 3``, you also need to have ``AvailableMenus = `` (empty). Still, it'd be better to not overwrite valid INI options that were edited manually by simply opening the launcher, in my opinion.

**Reply by charlie-in-spain** [reply]
Hi!

Ok, I'll wait for the next release then to get the Zidane's pose issue fixed, thanks! I thought it was something related to my mod. 

Regarding Memoria.ini bug, yes, I'd also added "AvailableMenus = "Equip", "SupportingAbility", but "AccessMenus" keeps reseting to "0" when opening the launcher. In the Memoria.ini file of my mod I've not included "AvailableMenus" and the access to the menu works though...

I don't quite get what you explain at the end, sorry, what do you mean by not overwriting INI options edited manually? In order to play with access to the menu it is necessary to add this parameter in Memoria.ini file, isn't it?

**Reply by Tirlititi** [reply]
I meant "it would be better if the launcher didn't automatically overwrite valid INI options that were edited manually when it gets opened".

**Reply by charlie-in-spain** [reply]
Then, how could I prevent this INI option from being overwritten? With parameter "NoAutoTrance", for example, it's working fine when setting it to "1". I've checked that "Enabled" is also set to "1" in [Battle] section. 

EDITED: In your Alternate Fantasy mod, these parameters are not overwritten, but I cannot access to the menu in battle even if they are added in your Memoria.ini file. I can access to it if I activate Auto-battle mod, quite strange. "NoAutoTrance" also works fine if set to "1" though, like in the original game.

**Comment by charlie-in-spain** [comment]
Sorry, I have another issue now... I've launched the game and I get this error: 

![launcher](https://github.com/user-attachments/assets/494408e1-e320-4c0c-90de-02a4e7c3fd71)

I've restaured the files from the game (the option available in Properties / Installed files in Steam), and then I repatched with the last Memoria update, but I still get that error and I can't launch the game. Before doing the repatch the game launches normally, but I get this error after repatching. 
Do you know why this happens? I hadn't changed anything since last time I played the game. 

Thanks!

**Reply by charlie-in-spain** [reply]
Hi again, 

I tried to repatch with the previous Memoria patch and the game launched correctly. Afterwards I repatched again with the last Memoria patch and then it also worked... I don't know what could have caused the error.

**Reply by charlie-in-spain** [reply]
Hello, 

I've got this issue again. Yesterday I could play the game normally from the launcher, and today I get the same error message. It looks like the launcher PLAY button doesn't work, so it's not possible to play the game. 
Have you ever experienced this issue?

**Comment by charlie-in-spain** [comment]
Hello! 

I've experienced a strange bug when using "WinItems" in Battlepatch.txt for Tantarian battle. It seems that "StealItems" works fine, but if I set 4 items for "WinItems" it works randomly. 
It's quite strange since I've tried to set no items in HW for this battle, but when it doesn't work the game makes you get the running shoes, which is the original drop item set for this. 
I've managed it to work if I set a different MAXHP though. For example, for only 20 HP it gets the correct items set in "WinItems" but if I change it to 2000 or 20000 it keeps dropping the running shoes. 

I've tried to transfer this battle to another enemy ID "slot" in HW, copying the enemy, spells, scripts and stuff, but the battle works the same way, keeping this bug. I wonder if this could be related to Memoria, the HW project or just this enemy in particular.

Regards!


EDIT: It seems it's due to the HW project. The mod I'm making is based on Tirlititi's HWS file, and I get the same result with his project. But if I try to export the project from the original game it works fine. Maybe it got somehow corrupted when planning the extra battle with the other 3 enemies in Tantarian battle?

**Reply by Tirlititi** [reply]
That's because you use Alternate Fantasy as a basis for your mod. Alternate Fantasy changes Tantarian's drops through its AI.

**Reply by charlie-in-spain** [reply]
Oh I see, I'll leave Tantarian's drops as they are then. Thanks for the info!
